tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/LogsInjection/LoggerImplWriteIntegration.cs

```diff
@@ -0,0 +1,87 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+using Datadog.Trace.ClrProfiler.CallTarget;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.NLog.LogsInjection
+{
+    /// <summary>
+    /// LoggerImpl.Write calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "NLog",
+        TypeName = "NLog.LoggerImpl",
+        MethodName = "Write",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { ClrNames.Type, "NLog.Internal.TargetWithFilterChain", "NLog.LogEventInfo", "NLog.LogFactory" },
+        MinimumVersion = "2.1.0",
+        MaximumVersion = "4.*.*",
+        IntegrationName = "NLog")]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class LoggerImplWriteIntegration
+    {
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TTargets">The type of the TargetWithFilterChain </typeparam>
+        /// <typeparam name="TLogEventInfo">The type of the LogEventInfo</typeparam>
+        /// <typeparam name="TLogFactory">The type of the LogFactory</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="loggerType">The instance of the logger type</param>
+        /// <param name="targetsForLevel">The instance of the targets for the level</param>
+        /// <param name="logEvent">The logging event instance</param>
+        /// <param name="factory">The LogFactory instance</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TTargets, TLogEventInfo, TLogFactory>(TTarget instance, Type loggerType, TTargets targetsForLevel, TLogEventInfo logEvent, TLogFactory factory)
+        {
+            var tracer = Tracer.Instance;
+
+            if (tracer.Settings.LogsInjectionEnabled)
+            {
+                if (DiagnosticContextHelper.Cache<TTarget>.Mdlc is { } mdlc)
+                {
+                    var state = DiagnosticContextHelper.SetMdlcState(mdlc, tracer);
+                    return new CallTargetState(scope: null, state);
+                }
+
+                if (DiagnosticContextHelper.Cache<TTarget>.Mdc is { } mdc)
+                {
+                    var removeSpanId = DiagnosticContextHelper.SetMdcState(mdc, tracer);
+                    return new CallTargetState(scope: null, removeSpanId);
+                }
+            }
+
+            return new CallTargetState(scope: null, state: null);
```

Comment:
I think there's a small performance benefit to using this?

```suggestion
            return CallTargetState.GetDefault();
```

Reply:
Either way, it's best practice 👍

Reply:
Changed in 8e280fbaeb36d290b1788ef949d74e4671193ff6